packages/asset-l2/contracts/Asset.sol

```diff
@@ -0,0 +1,453 @@
+                true,
+                revealHashes[i]
+            );
+            amounts[i] = 1;
+            unchecked {
+                i++;
+            }
+        }
+
+        _mintBatch(recipient, tokenIds, amounts, "");
+    }
+
+    /// @notice Special mint function for the bridge contract to mint assets originally created on L1
+    /// @dev Only the special minter role can call this function
+    /// @dev This function skips the catalyst burn step
+    /// @dev Bridge should be able to mint more copies of the same asset
+    /// @param originalTokenId The original token id of the asset
+    /// @param amount The amount of assets to mint
+    /// @param tier The tier of the catalysts to burn
+    /// @param recipient The recipient of the asset
+    /// @param revealed Whether the asset is to be minted as already revealed
+    /// @param revealHash The hash of the reveal
+    function bridgeMint(
+        uint256 originalTokenId,
+        uint256 amount,
+        uint8 tier,
+        address recipient,
+        bool revealed,
+        uint40 revealHash
+    ) external onlyRole(BRIDGE_MINTER_ROLE) {
+        // extract creator address from the last 160 bits of the original token id
+        address originalCreator = address(uint160(originalTokenId));
+        // extract isNFT from 1 bit after the creator address
+        bool isNFT = (originalTokenId >> 95) & 1 == 1;
+        require(amount > 0, "Amount must be > 0");
+        if (isNFT) {
+            require(amount == 1, "Amount must be 1 for NFTs");
+        }
+        // check if this asset has been bridged before to make sure that we increase the copies count for the same assers rather than minting a new one
+        // we also do this to avoid a clash between bridged asset nonces and non-bridged asset nonces
+        if (bridgedTokensNonces[originalTokenId] == 0) {
+            // increment nonce
+            unchecked {
+                creatorNonces[originalCreator]++;
+            }
+            // get current creator nonce
+            uint16 nonce = creatorNonces[originalCreator];
+
+            // store the nonce
+            bridgedTokensNonces[originalTokenId] = nonce;
+        }
+
+        uint256 id = generateTokenId(
+            originalCreator,
+            tier,
+            bridgedTokensNonces[originalTokenId],
+            revealed,
+            revealHash
+        );
+        _mint(recipient, id, amount, "");
+    }
+
+    /// @notice Extract the catalyst by burning assets of the same tier
+    /// @param tokenIds the tokenIds of the assets to extract, must be of same tier
+    /// @param amounts the amount of each asset to extract catalyst from
+    /// @param catalystTier the catalyst tier to extract
+    /// @return amountOfCatalystExtracted the amount of catalyst extracted
+    function recycleBurn(
+        address recycler,
+        uint256[] calldata tokenIds,
+        uint256[] calldata amounts,
+        uint256 catalystTier
+    )
```

Comment:
Same idea, all the recycling stuffs could be centralized into one contract
that way, we don't have to touch the asset contract to upgrade the recycling features if needed